torchtitan/experiments/vlm/datasets/mm_datasets.py

```diff
@@ -0,0 +1,422 @@
+                        elif isinstance(img, str):
+                            img = Image.open(img)
+
+                        if img.mode != "RGB":
+                            img = img.convert("RGB")
+
+                        # Resize maintaining aspect ratio
+                        img = resize_image_by_patch_count(
+                            img,
+                            max_patch_per_image=self.max_patch_per_image,
+                            patch_size=self.patch_size,
+                            merge_size=self.merge_size,
+                        )
+
+                        # Convert to numpy array and rescale to [0, 1]
+                        img_array = np.array(img) / 255.0
+
+                        # Normalize with OpenAI CLIP mean/std
+                        mean = np.array([0.48145466, 0.4578275, 0.40821073])
+                        std = np.array([0.26862954, 0.26130258, 0.27577711])
+                        img_array = (img_array - mean) / std
+
+                        # Convert to tensor in NTHWC format (1, H, W, 3)
+                        img_tensor = torch.from_numpy(img_array).float()
+                        img_tensor = img_tensor.unsqueeze(0)  # Add time dimension
+
+                        # Calculate number of image tokens needed
+                        (
+                            num_tokens,
+                            add_row_image_token_after,
+                            _,
+                        ) = calculate_image_tokens(
+                            img, patch_size=self.patch_size, merge_size=self.merge_size
+                        )
+
+                        processed_images.append(img_tensor)
+                        processed_text += BEGIN_OF_IMAGE_TOKEN
+
+                        # Add image tokens with row separators following dataset_utils pattern
+                        image_tokens = []
+                        for token_idx in range(num_tokens):
+                            image_tokens.append(IMAGE_TOKEN)
+
+                        processed_text += "".join(image_tokens)
+                        processed_text += END_OF_IMAGE_TOKEN
+
+                    except Exception as e:
+                        logger.warning(f"Error processing image {i}: {e}")
+
+            if not processed_images:
+                return None
+
+            # Add EOS token and tokenize
+            processed_text = processed_text + END_OF_IMAGE_TOKEN
+            tokens = self._tokenizer.encode(processed_text)
+
+            input_ids = torch.tensor(tokens)
+            labels = torch.tensor(tokens)
+
+            if len(input_ids) > self.seq_len:
+                logger.warning(
+                    f"Skipping sample with length {len(input_ids)} greater than max_seq_len {self.seq_len}"
+                )
+                return None
+
+            # Get special token IDs just like in dataset_utils.py
+            def _get_special_token_id(token):
```

Comment:
This part (to line 305) can be moved out side of process_sample function, because it only need to run once, not every time processing a sample